File: Python/Medium/2300_Successful_Pairs_of_Spells_and_Potions/solution.py
class Solution:
    def successfulPairs(self, spells: list[int], potions: list[int], success: int) -> list[int]:
        output = []
        potions.sort()

        for i in range(len(spells)):
            count = 0

            p1 = 0
            p2 = len(potions) - 1

            while p1 <= p2:
                mid = (p2 + p1) // 2
                power = spells[i] * potions[mid]
                if power < success:
                    p1 = mid + 1
                else:
                    p2 = mid - 1
            
            count += len(potions) - p1
            output.append(count)
        return output        

File: Golang/Medium/2300_Successful_Pairs_of_Spells_and_Potions/solution.go
package successfulpairsofspellsandpotions

import "sort"

func successfulPairs(spells []int, potions []int, success int64) []int {
	output := []int{}
	sort.Ints(potions)

	for i, _ := range spells {
		count := 0

		p1 := 0
		p2 := len(potions) - 1

		for p1 <= p2 {
			mid := (p1 + p2) / 2
			power := int64(spells[i] * potions[mid])
			if power < success {
				p1 = mid + 1
			} else {
				p2 = mid - 1
			}
		}
		count += len(potions) - p1
		output = append(output, count)
	}
	return output
}
